orttraining/orttraining/core/optimizer/memory_alleviation.cc

```diff
@@ -0,0 +1,754 @@
+                                   boundary_op_order_in_topological_sort)) {
+          // Remove the edge only connecting to backward op.
+          output_edges.push_back(graph_utils::GraphEdge::CreateGraphEdge(*node, *it, false));
+        }
+      }
+
+      if (!output_edges.empty()) {
+        // Remove the output edges of the node first
+        graph_utils::GraphEdge::RemoveGraphEdges(graph, output_edges);
+
+        // Create connections between the replacement node and the outgoing nodes.
+        for (const auto& output_edge : output_edges) {
+          graph.RemoveConsumerNode(node->MutableOutputDefs()[output_index]->Name(), node);
+
+          // Add new edge connecting the input with the output nodes directly.
+          // This also updates the destination node's input node args
+          graph.AddEdge(replacement_node->Index(), output_edge.dst_node, static_cast<int>(output_index),
+                        output_edge.dst_arg_index);
+        }
+      }
+    }
+  }
+
+  return graph_is_modified;
+}
+
+Status MemoryAlleviation::ApplyImpl(Graph& graph, bool& modified, int /*graph_level*/, const logging::Logger& logger)
+    const {
+  LOGS(logger, VERBOSE) << "Memory alleviation config: " << memory_alleviation_config_ << ", probe level: "
+                        << static_cast<int>(level_);
+
+  InlinedHashMap<std::string, std::pair<bool, bool>> fw_op_output_arg_used_map;
+  InlinedHashMap<NodeIndex, size_t> node_index_to_its_order_in_topological_sort_map;
+  int64_t boundary_op_order_in_topological_sort = PrepareForTransformation(graph,
+                                                                           fw_op_output_arg_used_map,
+                                                                           node_index_to_its_order_in_topological_sort_map);
+  if (boundary_op_order_in_topological_sort < 0) {
+    LOGS(logger, VERBOSE) << "No boundary op found. Skip memory alleviation.";
+    return Status::OK();
+  }
+
+  InlinedHashMap<const Node*, InlinedVector<size_t>> candidate_output_args_map;
+  ORT_RETURN_IF_ERROR(GetStashedActivationCandidates(graph, fw_op_output_arg_used_map, candidate_output_args_map,
+                                                     logger));
+
+  AlleviationSubGraphStores recompute_subgraph_stores;
+  AlleviationSubGraphStores recompute_with_compromise_subgraph_stores;
+  GraphViewer graph_viewer(graph);
+  const auto& node_ids = graph_viewer.GetNodesInTopologicalOrder();
+
+  // The first pass - find the candidate subgraphs.
+  for (int i = static_cast<int>(node_ids.size()) - 1; i >= 0; --i) {
+    Node* p_node = graph.GetNode(node_ids[i]);
+    if (p_node == nullptr) {
+      continue;
+    }
+
+    if (candidate_output_args_map.find(p_node) == candidate_output_args_map.end()) {
+      continue;
+    }
+
+    bool can_compromise_stashed_activation = false;
+    IsNodeRecomputable(*p_node, fw_op_output_arg_used_map, node_index_to_its_order_in_topological_sort_map,
```

Comment:
`IsNodeRecomputable` output is not consumed.

Reply:
Okay then I changed the return value to void, really it is not needed here.